src/vm/threadsuspend.cpp

```diff
@@ -2970,6 +2970,7 @@ void Thread::PulseGCMode()
     if (PreemptiveGCDisabled() && CatchAtSafePoint())
     {
         EnablePreemptiveGC();
+        m_MarshalAlloc.ClearUnallocated();
```

Comment:
I would not expect for this to be attached to PulseGCMode. It looks pretty fragile.

Reply:
Yeah, I was trying to find a bit of code that would operate when the typeloader is not doing any work, that could drop in and clean up the excess allocated memory. This fits the bill IMO, but as you say isn't a very clean place. (I'm trying to avoid any cross-thread interaction, as this api is hot enough that taking locks is quite visible in the perf data). Do you have any other ideas as to a good spot to put the logic?

Reply:
> Of course, we could allocate a slab at type load time start, which would probably be ok, but that's not really any better than this approach, and will require more memory overall. 

There are two uses of `Thread::m_MarshalAlloc` today:
- Type loader
- Interop

They have different nature.

I think the type loader one should be on the same plan as JIT, and the two should be tuned together. Having type loader on the same plan as JIT should be much easier to reason about.

The interop one use should grow into a more general purpose managed frame-local allocator that is integrated with exception unwind, etc. Some of this is discussed in https://github.com/dotnet/csharplang/issues/1344#issuecomment-369440829.

Reply:
Maybe the memory allocation piece should be split and discussed in a separate PR.

Reply:
So, we don't use the allocator for interop at all. (Its name is misleading...) I've re-worked it into being a thread local stack allocated thing, which achieves equivalent/better performance wins without the complexity of a pulsegcmode interaction. However, its not really relevant to tie this to the JIT as the JIT executes far less often, but needs vastly more memory as it executes. The JIT slab allocator is designed to avoid VirtualAlloc/VirtualFree pairs inside of the OS memory allocator, and that simply isn't an issue with this allocator and its users (as the amount of memory is so much less.)